### PR TITLE
Add "no-invalid-this" rule

### PR DESCRIPTION
**Pull request by AndyMoreland**
This PR adds a rule that flags usage of the `this` keyword outside of class-body scopes.

Often I find myself refactoring code that looks like:

```ts
class Foo {
    private aMethod() {
        doWorkOn(this.someState.x);
    }
}
```

to pull `aMethod` out into a function that looks like:

```ts
function(someState) {
    doWorkOn(someState.x);
}
```

When I make this change by hand I find that I screw up a lot and accidentally write:

```ts
function(someState) {
    doWorkOn(this.someState.x); /* note `this` */
}
```

which does not cause a compile error because typescript unhelpfully infers `this` with type `any`.

I've found that `this` outside of a `class` in code that is using ES6-class syntax is almost always a mistake. I wouldn't turn this on by default for all users of tslint, but it'd be very useful in my case (and I imagine for others.)

### CODE REVIEW

File: src/language/walker/scopeAwareRuleWalker.ts [hunk 1]
```diff
@@ -80,6 +80,7 @@ export abstract class ScopeAwareRuleWalker<T> extends RuleWalker {
             || node.kind === ts.SyntaxKind.ArrowFunction
             || node.kind === ts.SyntaxKind.ParenthesizedExpression
             || node.kind === ts.SyntaxKind.ClassDeclaration
+            || node.kind === ts.SyntaxKind.ClassExpression
```
Comment (AndyMoreland) [inline src/language/walker/scopeAwareRuleWalker.ts]: I believe this is correct but would appreciate a second set of eyes here.

Comment (AndyMoreland) [inline src/language/walker/scopeAwareRuleWalker.ts]: This is for things like:

```
const A = class {
...
}
```

Comment (JKillian) [inline src/language/walker/scopeAwareRuleWalker.ts]: 👍 Looks right to me

File: src/rules/unboundFunctionThisRule.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+import * as Lint from "../lint";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static FAILURE_STRING = "disallowed \"this\" outside of class body" ;
```
Comment (JKillian) [inline src/rules/unboundFunctionThisRule.ts]: Can we make this a little clearer? Maybe: `the "this" keyword is disallowed outside of a class body`

**Comment by JKillian** [comment]
Looks good to me! Although I'm not sure how I feel about the rule name, any thoughts @adidahiya? Also,  tweaking the error message would be great

File: src/rules/unboundFunctionThisRule.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public createScope(node: ts.Node): { inClass: boolean } {
+        let isClassScope = node.kind === ts.SyntaxKind.ClassDeclaration || node.kind === ts.SyntaxKind.ClassExpression;
+
+        return {
+            inClass: isClassScope,
+        };
+    }
+
+    protected validateThisKeyword(node: ts.Node) {
+        if (!this.getAllScopes().some(scope => scope.inClass)) {
```
Comment (JKillian) [inline src/rules/unboundFunctionThisRule.ts]: Normally as a style guideline we surround arrow func arguments with parens:
`(scope) => scope.inClass`. Just makes them easier to distinguish visually. This should be a lint rule!

**Comment by adidahiya** [comment]
I like `no-unbound-this` for the rule name

**Comment by adidahiya** [comment]
oh, actually it's called `no-invalid-this` in ESLint... we should probably be consistent? http://eslint.org/docs/rules/no-invalid-this

**Comment by JKillian** [comment]
Let's go with that: `no-invalid-this`

**Comment by myitcv** [comment]
Just a quick question on what we're looking to achieve here: aren't we ultimately trying to catch situations where the type of `this` is `any`?

Therefore I'm not totally clear how scopes etc comes into it.... other than, in the absence of full type information (#680), we're trying to look at scopes and infer that the type of `this` would be `any`.

Does that make sense?

**Comment by adidahiya** [comment]
@myitcv I'd consider this one of the stricter rules that I don't expect all TSLint users to enable. Yes, I would love to have full type information and check if `this` is `any`, but banning `this` outside `class` scopes is a decent analog that we can accomplish syntactically.

**Comment by AndyMoreland** [comment]
@myitcv when/if tslint gives me type info I'd be happy to write that rule.

**Comment by AndyMoreland** [comment]
I believe I've addressed the comments. I renamed to no-invalid-this and fixed the `scope =>` to be `(scope) =>`.

**Comment by adidahiya** [comment]
yeah we can merge this / refine the error message afterwards

File: src/tsconfig.json [hunk 4]
```diff
@@ -113,6 +113,7 @@
         "rules/tripleEqualsRule.ts",
         "rules/typedefRule.ts",
         "rules/typedefWhitespaceRule.ts",
+        "rules/unboundFunctionThisRule.ts",
```
Comment (JKillian) [inline src/tsconfig.json]: seems like the wrong file name here. If you run grunt locally it should automatically fix this

File: src/rules/noInvalidThisRule.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static FAILURE_STRING = "the \"this\" keyword is disallowed outside of a class body" ;
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new UnboundFunctionThisWalker(sourceFile, this.getOptions()));
+    }
+}
+
+class UnboundFunctionThisWalker extends Lint.ScopeAwareRuleWalker<{inClass: boolean}> {
```
Comment (JKillian) [inline src/rules/noInvalidThisRule.ts]: Class name is also not updated

File: test/tsconfig.json [hunk 6]
```diff
@@ -141,4 +142,4 @@
         "rule-tester/testData.ts",
         "rule-tester/utilsTests.ts"
     ]
-}
+}
```
Comment (JKillian) [inline test/tsconfig.json]: no reason to remove this newline 😉

File: src/rules/noInvalidThisRule.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    public createScope(node: ts.Node): { inClass: boolean } {
+        let isClassScope = node.kind === ts.SyntaxKind.ClassDeclaration || node.kind === ts.SyntaxKind.ClassExpression;
```
Comment (JKillian) [inline src/rules/noInvalidThisRule.ts]: can be a const

File: src/rules/noInvalidThisRule.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
```
Comment (JKillian) [inline src/rules/noInvalidThisRule.ts]: Can be 2016

**Comment by JKillian** [comment]
Still needs a few tweaks, then it should be good to merge

Note to self: add documentation / tweak error message after this merges

**Comment by AndyMoreland** [comment]
I've made the four changes you requested. Sorry for the sloppiness!

**Comment by JKillian** [comment]
Sweet, merged! Release should be sometime this week, Thursday at the latest

**Comment by AndyMoreland** [comment]
Thanks!

**Comment by aaronresnick** [comment]
> @myitcv when/if tslint gives me type info I'd be happy to write that rule.

@AndyMoreland I believe writing a rule that checks for `this` usage only when its type is any is possible now, right? If so, that would be super helpful for me. Are you still interested in writing it?

**Comment by dugagjinll** [comment]
```````

**Comment by goktugyil** [comment]
In mocha, we need to use this.skip outside the scope of classes: http://prntscr.com/ns01dr
This gives me a warning. Is there a way to disable cases like these? https://mochajs.org/

**Pull request merged**